libcaf_core/caf/flow/step/take_last.hpp

```diff
@@ -0,0 +1,55 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/fwd.hpp"
+
+#include <cstddef>
+
+namespace caf::flow::step {
+
+template <class T>
+class take_last {
+public:
+  using input_type = T;
+
+  using output_type = T;
+
+  explicit take_last(size_t num) : last_elements_count_(num) {
+    // nop
+  }
+
+  take_last(take_last&&) = default;
+  take_last(const take_last&) = default;
+  take_last& operator=(take_last&&) = default;
+  take_last& operator=(const take_last&) = default;
+
+  template <class Next, class... Steps>
+  bool on_next(const input_type& item, Next& next, Steps&... steps) {
+    elements_.emplace_back(std::move(item));
```

Comment:
This implementation has multiple issues. If you only take `m` elements, this still requires to store all `n` elements. This is a total waste of both memory and CPU cycles (due to unnecessary copying operations for frequent re-allocations).

This operator should not buffer more than `m` elements at any point in time. It also shouldn't need to allocate memory more than once, because we know the maximum space requirement at construction time and can use a simple ring buffer.

Reply:
I had planned this implementation but I came across this doc in ReactiveX (http://introtorx.com/Content/v1.0.10621.0/05_Filtering.html#SkipLastTakeLast) and decided that we need to cache all the values. For ring buffer I was planning to use boost circular buffer https://www.boost.org/doc/libs/1_81_0/doc/html/circular_buffer.html. But, as we currently don't use anything boost the closest thing would be std::deque (https://en.cppreference.com/w/cpp/container/deque) which still has some performance issues. Can you tell me your thoughts on this?

Reply:
Pulling in a new dependency for this would be overkill. All we need here is an index for the current writing position, a bit of module math and a flag that tells us whether we've filled the cache at least once (to know whether we `< n` elements).

Reply:
In that case, I think that it is better to create a class for ring buffer data structure instead of just adding the code directly into the observable class in case we need it for other classes in future. What are your thoughts on this?

Reply:
Yeah, good point. We already have a ring buffer class, but that is for synchronized producers/consumers. I've filed #1508 to give that old class a better name. When that's merged, you can go ahead an implement a new class `detail::ring_buffer` for this use case. But keep the interface simple. Probably all we need is `push_back`, `size`, `front` and `pop_front`.